File: epidemic_model.py
import math
import numpy as np
import random
#import matplotlib.pyplot as plt

class Canvas:

    def __init__(self, height, width, framerate, secs_per_day):
        self.height = height
        self.width = width
        self.framerate = framerate
        self.sec_per_day = secs_per_day
        self.frames_per_day = framerate*secs_per_day
        self.time = 0

    def time_step(self):
        self.time += 1

class Bubble:

    def __init__(self, x, y, r, particles, quarantine=False):
        self.x = x
        self.y = y
        self.r = r
        self.particles = particles
        self.quarantine = quarantine
    
    def check_infection(self):
        quarantine = False
        for p in self.particles:
            if p.state == 'I':
                quarantine = True
        return quarantine

class Particle:

    def __init__(self,x, y, r, heading, speed, state, 
                 alpha, gamma, canvas, bubble):
        self.x = x
        self.y = y
        self.r = r
        self.heading = heading
        self.speed = speed
        self.state = state
        self.state_time = 0
        self.canvas = canvas
        self.bubble = bubble
        ## get cartesian speed
        self.speedx = self.speed*math.cos(self.heading)
        self.speedy = self.speed*math.sin(self.heading)
        ## get state transitions
        self.latent_period = alpha*canvas.frames_per_day
        self.infectious_period = gamma*canvas.frames_per_day

    def move(self, bubbles):
        # add a wiggle
        self.heading += random.uniform(-0.5, 0.5)
        # avoid any quarantined bubbles
        other_bubbles = [b for b in bubbles if b != self.bubble]
        for b in other_bubbles:
            if b.quarantine:
                # chenge direction to avoid bubble
                self.heading = self.avoid_bubble(b)
        # turn if hitting the side
        self.heading = self.bounce_at_edge()
        # adjust so 0 < heading < 2*pi
        if self.heading < 0:
            self.heading += 2*math.pi
        elif self.heading > 2*math.pi:
            self.heading -= 2*math.pi
        # update the location based on the speed
        self.update_location()
    
    def update_location(self):
        self.speedx = self.speed*math.cos(self.heading)
        self.speedy = self.speed*math.sin(self.heading)
        self.x += self.speedx
        self.y += self.speedy

    def bounce_at_edge(self):
        new_heading = self.heading
        if (self.x < 0) or (self.x > self.canvas.width):
            new_heading = math.pi - self.heading
        if (self.y < 0) or (self.y > self.canvas.height):
            new_heading = 2*math.pi - self.heading
        return new_heading         

    def avoid_bubble(self, b):
        # check distance
        d = math.dist((self.x, self.y), (b.x, b.y))
        min_distance = self.r + b.r + 50
        # initialis the new angle as the current angle
        new_angle = self.heading
        if d<min_distance:
            # calculate reflection angle
            angle_to_b = math.atan2(b.y-self.y, b.x-self.x)  
            new_angle = -2*angle_to_b - self.heading
        return new_angle

    def go_home(self):
        # add a random wiggle
        #self.heading += random.uniform(-0.1, 0.1)
        target_heading = math.atan2(self.bubble.y - self.y, self.bubble.x- self.x)
        dist = math.dist((self.x, self.y), (self.bubble.x, self.bubble.y))
        if dist > self.bubble.r:
            angle_diff = target_heading - self.heading
            if angle_diff > math.pi:
                angle_diff -= 2*math.pi
            elif angle_diff < -2*math.pi:
                angle_diff += 2*math.pi
            # multiply by 0.2 to smooth motion
            self.heading += angle_diff*0.2

    def infect_particles(self, particles):
        if self.state != 'I':
            pass
        for p in particles:
            dist = math.dist((self.x, self.y), (p.x, p.y))
            if (dist < self.r) and (p.state == 'S'):
                p.state = 'E'

    def transition_state(self):
        if(self.state=='E') and (self.state_time > self.latent_period):
            self.state = 'I'
            self.state_time = 0
        if(self.state=='I') and (self.state_time > self.infectious_period):
            self.state = 'R'
            self.state_time = 0
    
    def time_step(self):
        self.state_time +=1

def map_value(n, start1, stop1, start2, stop2):
    return (n - start1) / (stop1 - start1) * (stop2 - start2) + start2

if __name__ == '__main__':
    n_bubbles = 12
    canvas = Canvas(height=800, width=800,framerate=60,secs_per_day=12)
    cols = np.floor(math.sqrt(n_bubbles))
    rows = np.floor(n_bubbles/cols)
    particles = []
    bubbles = []
    for i in range(n_bubbles):
        col = i%cols
        row = i//cols
        x = map_value(col, 0, cols-1, 100, canvas.width-100)
        y = map_value(row, 0, rows-1, 100, canvas.height-100)
        bubbles.append(Bubble(x, y, 20, particles, False))
    
    n_particles = canvas.width//10
    
    for i in range(n_particles):
        bubble = random.choice(bubbles)
        p = Particle(bubble = bubble,
                     x = bubble.x + random.uniform(-10,10),
                     y = bubble.y + random.uniform(-10,10),
                     heading = random.uniform(0, 2*math.pi),
                     speed = random.uniform(1,3),
                     r = 20,
                     canvas=canvas,
                     state = 'S',
                     alpha = 1,
                     gamma = 2)
        particles.append(p)

    # add an exposed particle
    bubble = bubbles[0]
    p = Particle(bubble = bubble,
                 x = bubble.x + random.uniform(-10,10),
                 y = bubble.y + random.uniform(-10,10),
                 heading = random.uniform(0, 2*math.pi),
                 speed = random.uniform(1,3),
                 r = 20,
                 canvas=canvas,
                 state = 'E',
                 alpha = 0.75,
                 gamma = 2)

    epi_curve = []
    n_steps = 6000
    for i in range(6000):
        for bubble in bubbles:
            bubble.quarantine = bubble.check_infection()

        for particle in particles:
            particle.move(bubbles)
            particle.transition_state()
            if particle.state=='I':
                particle.infect_particle()
            particle.time_step()
            if canvas.time%canvas.frames_per_day < canvas.frames_per_day*2/3:
                particle.go_home()
            elif particle.bubble.quarantine:
                particle.go_home()

        n_infected = sum([p for p in particles if p.state == 'I'])
        epi_curve.append(n_infected)
        canvas.time_step
        
    




            

File: epidemic_model.js
// this class describes the properties of a single particle.
function getShape(mu, sigma){
  let x1 = mu**2
  let x2 = Math.sqrt(x1 + sigma**2)
  return Math.log(x1/x2)
}

function getScale(mu, sigma){
  let x1 = 1 + (mu**2)/(sigma**2)
  return Math.sqrt(Math.log(x1))
}

function logNormal(mu, sigma) {

  let muHat = getShape(mu, sigma)
  let sigmaHat = getScale(mu, sigma)
  let xi1 = random(0,1)
  let xi2 = random(0,1)
  let z1 = Math.sqrt(-2*Math.log(1.0-xi1))
  let z2 = Math.sin(2*Math.PI*xi2)
  return Math.exp(muHat * sigmaHat*z1*z2) 
}


let muLatent = 1
let sigmaLatent = 0.5
let muInfectious = 2
let sigmaInfectious = 1
let framesPerSec = 60
let daySeconds = 9
let speedMin = 2
let speedMax = 4
let dayFrames = framesPerSec*daySeconds


class Particle {
    // setting the co-ordinates, radius and the
    // speed of a particle in both the co-ordinates axes.
      constructor(){
        //console.log(logNormal(10,2))
        this.x = random(0,width);
        this.y = random(0,height);
        this.r = 20;
        this.heading = random(0, 2*Math.PI)
        this.speed = random(speedMin,speedMax)
        this.xSpeed = this.speed*Math.cos(this.heading);
        this.ySpeed = this.speed*Math.sin(this.heading);
        this.state = "S"
        this.stateTime = 0
        this.bubble = 0
        this.latentPeriod = logNormal(muLatent, sigmaLatent)*dayFrames
        this.infectiousPeriod = logNormal(muInfectious, sigmaInfectious)*dayFrames
        //this.latentPeriod = muLatent*dayFrames
        //this.infectiousPeriod = muInfectious*dayFrames
      }

    
    // creation of a particle.
      createParticle() {
        noStroke();
        if(this.state=="S")
            fill('rgba(200,200,200,0.8)');
            
        else if(this.state=="E")
            fill('rgba(255, 165, 0, 0.8)');

        else if(this.state=="I")
            fill('rgba(255, 0, 0, 0.8)');

        else if(this.state=="R")
            fill('rgba(0, 128, 128, 0.8)');

        circle(this.x,this.y,this.r);
      }
    
    // setting the particle in motion.
      moveParticle(bubbles) {        

        // turn if hitting another bubble
        let idx = bubbles.indexOf(this.bubble)
        let otherBubbles = bubbles.slice(0, idx).concat(bubbles.slice(idx + 1))
        otherBubbles.forEach(b=>{
          if(b.quarantine==true){
            let d = dist(this.x, this.y, b.x, b.y);
            let minDistance = this.r + b.r + 50;
          
            if (d < minDistance) {
              // Adjust the particle's heading based on the collision with the other bubble
              let angleToBubble = atan2(b.y - this.y, b.x - this.x);
              this.heading = -2*angleToBubble - this.heading;
            }
        }
        }) 

        // turn if hitting the side
        if(this.x < 0 || this.x > width){
          this.heading = Math.PI - this.heading;    
          }
        if(this.y < 0 || this.y > height){
          this.heading = 2*Math.PI - this.heading;
        }
        
        if(this.heading > 2*Math.PI){
          this.heading -=2*Math.PI
        }
        if(this.heading <0){
          this.heading += 2*Math.PI
        }
        
        this.xSpeed = this.speed*Math.cos(this.heading)
        this.ySpeed = this.speed*Math.sin(this.heading)
        this.x+=this.xSpeed;
        this.y+=this.ySpeed;
      }

    
    goHome(){
      this.heading += random(-0.1,0.1)
      //console.log("Bubble properties:", this.bubble)
      let targetHeading = Math.atan2((this.bubble.y-this.y),(this.bubble.x-this.x))
      let dis = dist(this.x, this.y, this.bubble.x, this.bubble.y)
      //stroke('rgba(255,255,255,0.5)');
      //line(this.x,this.y,this.bubble.x,this.bubble.y);          
      if (dis > this.bubble.r) {
        // Gradually adjust heading towards targetHeading
        let angleDifference = targetHeading - this.heading;
        if (angleDifference > Math.PI) {
          angleDifference -= 2 * Math.PI;
        }
        else if (angleDifference < -Math.PI) {
          angleDifference += 2 * Math.PI;
        }
        this.heading += angleDifference * 0.2; // Adjust the factor for smoother transition
      }
    }
    
    // quarantine infected particles
      quarantineInfected(){
        if (this.state=="I"){
            this.xSpeed = 0
            this.ySpeed = 0
        }
      }
    // between particles which are less than a certain distance apart
      infectParticle(particles) {
        particles.forEach(element =>{
          let dis = dist(this.x,this.y,element.x,element.y);
          if (this.state=="S" & element.state=="I" & dis <this.r){
              this.state='E';
              this.stateTime=0;
          }
            })
        }
    
      transitionState(){
        if(this.state==="E" & this.stateTime > this.latentPeriod){
            this.state="I";
            this.stateTime=0;
        }
        if(this.state==="I" & this.stateTime > this.infectiousPeriod){
            this.state="R";
            this.stateTime=0;
            //this.xSpeed=random(-1,1)
            //this.ySpeed=random(-1,1)
        }
      }
    // this function creates the connections(lines)
    // between particles which are less than a certain distance apart
      joinParticles(particles) {
        particles.forEach(element =>{
          let dis = dist(this.x,this.y,element.x,element.y);
          if(dis<this.r + element.r) {
            stroke('rgba(255,255,255,0.5)');
            line(this.x,this.y,element.x,element.y);
          }
        });
      }
    

      updateTimer() {
        this.stateTime+=1;
      }
}
class Bubble {

  constructor(){
    this.time = 0
    this.x = random(0, width)
    this.y = random(0, height)
    this.r = 20
    this.particles = []
    this.quarantine = false
    }
  createBubble(){
    noFill()
    if (this.quarantine == false){
      stroke('rgba(200,200,200,0.6)');
    }
    else{
      stroke('rgba(200,100,100,0.6)');
    }
    
    strokeWeight(3)
    circle(this.x, this.y, 100);

  }

  checkInfection(){
    this.quarantine= false
    this.particles.forEach(p =>{
    if(p.state=="I"){
      this.quarantine = true
      }

    })


    }
  

    updateTimer(){
      this.time +=1;
    
    }
    }



    // an array to add multiple particles
    let particles = [];
    let bubbles = [];
    let nBubbles = 12


    
    function setup() {
      createCanvas(750,750);

      let cols = int(sqrt(nBubbles));  // Number of columns (adjust as needed)
      let rows = ceil(nBubbles / cols); // Number of rows
    

      
      for (let i = 0; i < nBubbles; i++) {
        let col = i % cols;
        let row = floor(i / cols);
    
        let x = map(col, 0, cols - 1, 100, width - 100);
        let y = map(row, 0, rows - 1, 100, height - 100);
    
        let b = new Bubble();
        b.x = x;
        b.y = y;
        bubbles.push(b);
      }

      for(let i = 0;i<width/10;i++){
        let p = new Particle();
        p.bubble = bubbles[Math.floor(random(bubbles.length))];;
        p.x = p.bubble.x + random(-10, 10);
        p.y = p.bubble.y + random(-10, 10);
        p.bubble.particles.push(p);
        particles.push(p);
      }

      // add a single infected particle
      pExposed = new Particle();
      pExposed.state = "E";
      pExposed.bubble = bubbles[0]
      pExposed.x = pExposed.bubble.x + random(-10,10)
      pExposed.y = pExposed.bubble.y + random(-10,10)
      pExposed.latentPeriod = (daySeconds-1)*framesPerSec
      console.log(pExposed.latentPeriod)
      pExposed.bubble.particles.push(pExposed)
      particles.push(pExposed)
    //  for (let i=0; i< particles.length;i++){
    //   console.log(particles[i].bubble.x)
    //    //console.log(particles[i].bubble.y)
    //  }
      
    }

  
    function draw() {
      frameRate(framesPerSec);
      background('#0f0f0f');
      for(let i=0;i<bubbles.length;i++){
        bubbles[i].createBubble()
        //bubbles[i].checkInfection()
      }
      for(let i = 0;i<particles.length;i++) {
        particles[i].createParticle();
        particles[i].moveParticle(bubbles);
        particles[i].transitionState();
        particles[i].infectParticle(particles);
        particles[i].updateTimer();
        particles[i].joinParticles(particles.slice(i));
        if(frameCount%dayFrames <dayFrames*2/3){
          //particles[i].speed = 3
          particles[i].goHome();
            //background('rgba(100,0,100,0.01)');
        }
        else{
          //particles[i].speed = random(1,2)
          if(particles[i].bubble.quarantine){
            particles[i].goHome();
          }
        }

        
        //particles[i].quarantineInfected();


      }
    }
    